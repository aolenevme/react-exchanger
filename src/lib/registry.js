import store from "../store.js";

function updateIn(currentStore, payload) {
    const [path, value] = payload;

    path.reduce((acc, key, index) => {
        if (index === path.length - 1) {
            acc[key] = value;
        } else {
            return acc[key];
        }
    }, currentStore);
}

const storeEventsRegistry = {};

function dispatch(id, payload) {
    if (!id) {
        console.error("Enable to dispatch a store event without its id");

        return;
    }

    if (!storeEventsRegistry[id]) {
        console.error(
            `There isn\`t a registered handler for this store event id: ${id}`
        );

        return;
    }

    updateIn(store, storeEventsRegistry[id](store, payload));
}

function regEventStore(id, storeEvent) {
    if (!id) {
        console.error("Unable to register the event without its id");
    }

    if (!storeEvent) {
        console.error("Unable to register the store event without its handler");
    }

    if (storeEventsRegistry[id]) {
        console.error(
            `Unable to register the store event because it is already 
            registered: ${storeEventsRegistry[id]}`
        );
    }

    storeEventsRegistry[id] = storeEvent;
}

export {dispatch, regEventStore};
