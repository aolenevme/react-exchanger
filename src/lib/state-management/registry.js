import store from "../../store/store.js";

const storeEventsRegistry = {};

function dispatch(id, payload) {
    if (!id) {
        console.error("Enable to dispatch a store event without its id");

        return;
    }

    if (!storeEventsRegistry[id]) {
        console.error(
            `There isn\`t a registered handler for this store event id: ${id}`
        );

        return;
    }

    updateIn(store, storeEventsRegistry[id](store, payload));
}

function updateIn(currentStore, payload) {
    const [path, value] = payload;

    path.reduce((acc, key, index) => {
        if (index === path.length - 1) {
            acc[key] = value;
        } else {
            return acc[key];
        }
    }, currentStore);
}

function regEventStore(id, storeEvent) {
    if (!id) {
        console.error("Unable to register the event without its id");
    }

    if (!storeEvent) {
        console.error("Unable to register the store event without its handler");
    }

    if (storeEventsRegistry[id]) {
        console.error(
            `Unable to register the store event because it is already 
            registered: ${storeEventsRegistry[id]}`
        );
    }

    storeEventsRegistry[id] = storeEvent;
}

export {dispatch, regEventStore};
