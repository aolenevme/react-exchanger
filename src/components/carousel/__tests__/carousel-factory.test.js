import React from "react";
import renderer from "react-test-renderer";
import constant from "lodash/constant.js";

import * as registry from "../../../lib/state-management/registry.js";
import CarouselFactory, {inputFactory} from "../carousel-factory.js";
import MUTATE_STORE from "../../../events/mutate-store.js";

describe("<CarouselFactory />", () => {
    it("is rendered correctly", () => {
        expect(renderer
            .create(<CarouselFactory />).toJSON())
            .toMatchSnapshot();

        expect(renderer
            // eslint-disable-next-line max-len
            .create(<CarouselFactory getSpecification={constant({inputSign: "-", isDisabled: false, getRate: constant(1)})}/>).toJSON())
            .toMatchSnapshot();
    });

    describe("inputFactory", () => {
        it("creates <Input /> that updates store", () => {
            registry.dispatch = jest.fn();

            const testInputEvent = {target: {value: 123}};

            const Input = inputFactory();

            renderer.create(Input).toJSON().children[1].props.onInput(testInputEvent);

            // eslint-disable-next-line prefer-destructuring
            const [dispatchId, mutationEvent] = registry.dispatch.mock.calls[0];

            expect(dispatchId).toEqual(MUTATE_STORE);
            expect(mutationEvent()).toEqual([["exchangeAmount"], testInputEvent.target.value]);
        });
    });
});
