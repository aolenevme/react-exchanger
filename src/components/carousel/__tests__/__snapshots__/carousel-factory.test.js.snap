// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarouselFactory /> is rendered correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  overflow: hidden;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding-bottom: 2rem;
}

.c2 {
  display: inline-block;
  margin: 0 0.142rem;
  width: 0.5rem;
  height: 0.5rem;
  border-radius: 50%;
  background-color: #f3f6ff;
  -webkit-transition: background-color 0.1s ease 0s;
  transition: background-color 0.1s ease 0s;
}

.c3 {
  display: inline-block;
  margin: 0 0.142rem;
  width: 0.5rem;
  height: 0.5rem;
  border-radius: 50%;
  background-color: #5591f7;
  -webkit-transition: background-color 0.1s ease 0s;
  transition: background-color 0.1s ease 0s;
}

<div>
  <div
    className="c0"
    onMouseDown={[Function]}
    onMouseUp={[Function]}
  />
  <div
    className="c1"
  >
    <span
      className="c2"
    />
    <span
      className="c3"
    />
    <span
      className="c3"
    />
  </div>
</div>
`;

exports[`<CarouselFactory /> is rendered correctly 2`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  overflow: hidden;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding-bottom: 2rem;
}

.c2 {
  display: inline-block;
  margin: 0 0.142rem;
  width: 0.5rem;
  height: 0.5rem;
  border-radius: 50%;
  background-color: #f3f6ff;
  -webkit-transition: background-color 0.1s ease 0s;
  transition: background-color 0.1s ease 0s;
}

.c3 {
  display: inline-block;
  margin: 0 0.142rem;
  width: 0.5rem;
  height: 0.5rem;
  border-radius: 50%;
  background-color: #5591f7;
  -webkit-transition: background-color 0.1s ease 0s;
  transition: background-color 0.1s ease 0s;
}

<div>
  <div
    className="c0"
    onMouseDown={[Function]}
    onMouseUp={[Function]}
  />
  <div
    className="c1"
  >
    <span
      className="c2"
    />
    <span
      className="c3"
    />
    <span
      className="c3"
    />
  </div>
</div>
`;
